### ISSUE TITLE
Model calibration and replacement nodes

### ISSUE THREAD

**Issue by iMarang** [APSIMInitiative/ApsimX#5384]
I’m calibrating the APSIM model (next generation version) for the genetic coefficients of an EGAGregory derivative cultivar To do this calibration, I’m varying the coefficients to find the optimal set that minimises the difference in simulated vs observed phenology, measured as Zadoks stage.  I’ve written code to do this in Python, which:
 
1. Creates a unique .apsimx input file with the fertilizer level
2. Modifies the Wheat.json (lines 7026-7033) to replace the ‘FixedValue’ for each coefficient with realistic values (a permutation set of the range based on the full parameter values for all Australian cultivars)
3. Runs the model, extracts the .db output into .csv and compiles all results
4. Compares the Zadok stages Days After Sowing of 4 observed dates vs simulated dates to get a mean error for each coefficient set

I’ve found that regardless of the values of these 4 coefficients, the result is exactly the same Days After Sowing error.  When I use the report variables to show the FixedValue used it’s always the same set.  I’ve tried adjusting these by varying the model constants, however, this has the same result and shows the same original FixedValue in the report.
 
I’ve previously been in contact with an APSIM developer and he mentioned that the code is Read-Only within Models.exe to prevent people breaking the models accidentally, though we were talking generally and not about the genetic coefficients.  I’ve tried creating a Replacement Node to address this, however that is running into an error where it isn’t creating the required ‘_Checkpoint’ table in the .db (this is backed up by the output .db being only 4K, whereas a complete run is ~560KB). Attached is a zipped example .apsimx file from the replacement node. 

[working0_exp1_150_MED_e0_5-0_1-2_5-0_105.apsimx.zip](https://github.com/APSIMInitiative/ApsimX/files/4906704/working0_exp1_150_MED_e0_5-0_1-2_5-0_105.apsimx.zip)

I hope that makes sense.  Can you see where I’m going wrong?  I’d appreciate any ideas you have on how to go about the genetic coefficient calibration

**Comment by BrianCollinss** [comment]
Dear @iMarang you really don't need most of the text to explain the problem. Try to be brief so developers have time to read and respond.

Please attach the APSIM file, if possible.

**Comment by iMarang** [comment]
Thanks @BrianCollinss, noted.  Edited for brevity and attached the file.

**Comment by BrianCollinss** [comment]
@iMarang what you need to do is to put your **Simulation** directly under the **Simulations** folder along with the **Replacements**. If you need to have an editable copy of **Wheat** model under **Replacements**, open **Wheat.json** in **ApsimX\Models\Resources** and copy the content, then go back to the GUI and paste it under **Replacements**. Whatever you put under **Replacements** will _replace_ the object with the same name in the **Simulation**. This way you will have a copy of the wheat model without needing to modify the original version. After this, you need to make sure you play with Gregory parameters in your APSIM file.

**Comment by iMarang** [comment]
Perfect.  Thanks @BrianCollinss I'm now seeing variability across the parameters

**Comment by ashehad** [comment]
Hi @BrianCollinss - I have been trying to do a similar exercise as you suggested above for but AgPasture in APSIM NextGen. I have a clarification and a also question. The replacements is a 'folder' - is this correct? I don't know where to get the 'icon' of the AGPasture species for Ryegrass (similar to your suggestion) so that I can copy and paste all the parameter values of Ryegrass into it.  Thanks

**Comment by BrianCollinss** [comment]
Right click on `Simulations` and select Add Model and search for Replacements. 

I am not following your next question.

**Comment by ashehad** [comment]
Ok, Thanks. I was able to add the Replacements and the AGPRyegrass, which has these four children: 'Leaf', 'Stem', 'Stolon' and 'Root'. However, I need to look at for e.g. Gsmax. This parameter isn't corresponding to either of the four children - rather directly as "AGPRyegrass.Gsmax". 

I checked that in the case of Wheat, it is possible to look at Replacements.Wheat.Leaf.Gsmax. I see no issues here.
